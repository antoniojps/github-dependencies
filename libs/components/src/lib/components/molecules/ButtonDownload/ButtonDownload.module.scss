.options {
  display: inline-flex;
  flex-direction: column;
}

.option {
  display: inline-flex;
  align-items: center;
  padding: var(--space-2x) var(--space-5x);
  font-size: var(--size-xs);
  color: var(--yellow);
}

.name {
  border-bottom: 1px solid var(--yellow);

  input {
    background-color: var(--background);
    outline: 0;
    border: none;
    width: 100%;
    font-size: 12px;
    color: var(--yellow);
    background: transparent;
    border: none;
    outline: none;
    text-align: right;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    padding-left: var(--space);
  }
}

.buttonOptions {
  display: inline-flex;
  justify-content: space-between;
  border-bottom: 1px solid var(--yellow);

  &:last-of-type {
    border-bottom: none;
  }

  &Disabled {
    opacity: 0.5;
    background-color: transparent !important;

    &:hover {
      opacity: 1;
    }
  }

  button {
    border: none;
    outline: 0;
    transition: opacity 200ms;
    cursor: pointer;
    background-color: var(--background-light);
    text-transform: lowercase;
  }
}
