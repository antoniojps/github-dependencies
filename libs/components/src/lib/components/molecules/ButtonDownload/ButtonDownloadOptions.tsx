import React, { useReducer } from 'react';
import styles from './ButtonDownload.module.scss';
import classNames from 'classnames';
import { useDebounce } from 'react-use';
import { DownloadOptions } from '@github-graphs/types';

type ButtonDownloadOptionsProps = {
  initialOptions?: DownloadOptions;
  onChange: (options: DownloadOptions) => void;
};

export const defaultOptions: DownloadOptions = {
  name: '',
  size: '2x',
  format: 'png',
};

const SIZES: DownloadOptions['size'][] = ['1x', '2x'];
const FORMATS: DownloadOptions['format'][] = ['svg', 'jpeg', 'png'];

type Action =
  | { type: 'SET_NAME'; payload: string }
  | { type: 'SET_SIZE'; payload: DownloadOptions['size'] }
  | { type: 'SET_FORMAT'; payload: DownloadOptions['format'] };

function reducer(options: DownloadOptions, action: Action): DownloadOptions {
  switch (action.type) {
    case 'SET_NAME':
      return { ...options, name: action.payload };
    case 'SET_SIZE':
      return { ...options, size: action.payload };
    case 'SET_FORMAT':
      return { ...options, format: action.payload };
    default:
      throw new Error();
  }
}

export const ButtonDownloadOptions = ({ initialOptions, onChange }: ButtonDownloadOptionsProps) => {
  const [state, dispatch] = useReducer(reducer, initialOptions);

  useDebounce(
    () => {
      onChange(state);
    },
    200,
    [state]
  );

  return (
    <div className={styles.options}>
      <div className={`${styles.name} ${styles.option}`}>
        <label>name</label>
        <input
          type="text"
          placeholder="dependencies"
          name="name"
          value={state.name}
          onChange={(e) => dispatch({ type: 'SET_NAME', payload: e.target.value })}
        />
      </div>

      <div className={`${styles.buttonOptions} ${styles.option}`}>
        <label>size</label>
        <div className={styles.buttons}>
          {SIZES.map((size) => (
            <button
              key={size}
              className={classNames({ [styles.buttonOptionsDisabled]: state.size !== size })}
              onClick={() => dispatch({ type: 'SET_SIZE', payload: size })}
            >
              {size}
            </button>
          ))}
        </div>
      </div>

      <div className={`${styles.buttonOptions} ${styles.option}`}>
        <label>format</label>
        <div className={styles.buttons}>
          {FORMATS.map((format) => (
            <button
              key={format}
              className={classNames({ [styles.buttonOptionsDisabled]: state.format !== format })}
              onClick={() => dispatch({ type: 'SET_FORMAT', payload: format })}
            >
              {format}
            </button>
          ))}
        </div>
      </div>
    </div>
  );
};

ButtonDownloadOptions.defaultProps = {
  initialOptions: defaultOptions,
};
