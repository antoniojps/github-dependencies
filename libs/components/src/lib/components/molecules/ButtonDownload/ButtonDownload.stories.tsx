import React, { ReactElement } from 'react';
import { ButtonDownload } from './ButtonDownload';
import { action } from '@storybook/addon-actions';

export default {
  title: 'molecules/ButtonDownload',
};

export const Basic = (): ReactElement => {
  return (
    <ButtonDownload
      onClick={(options) => {
        action('onClick');
        console.log(options);
      }}
    />
  );
};
