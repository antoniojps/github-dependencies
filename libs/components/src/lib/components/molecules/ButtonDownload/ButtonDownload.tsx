import React, { useState } from 'react';
import { ButtonOptionsPopover } from '../ButtonOptionsPopover/ButtonOptionsPopover';
import { ButtonDownloadOptions, defaultOptions } from './ButtonDownloadOptions';
import { DownloadOptions } from '@github-graphs/types';

type ButtonDownloadProps = {
  onClick: (options: DownloadOptions) => void;
  options?: DownloadOptions;
};

export const ButtonDownload = ({ onClick, options = defaultOptions }: ButtonDownloadProps) => {
  const [downloadOptions, setDownloadOptions] = useState(options);

  return (
    <ButtonOptionsPopover
      onClick={() => onClick(downloadOptions)}
      content={<ButtonDownloadOptions onChange={setDownloadOptions} />}
    >
      Download
    </ButtonOptionsPopover>
  );
};
