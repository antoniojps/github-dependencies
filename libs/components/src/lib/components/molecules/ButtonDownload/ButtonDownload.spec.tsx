import React from 'react';
import { render } from '@testing-library/react';
import { noop } from 'lodash';

import { ButtonDownload } from './ButtonDownload';
import { ButtonDownloadOptions } from './ButtonDownloadOptions';

describe('ButtonDownload', () => {
  it('renders', () => {
    const { asFragment } = render(<ButtonDownload onClick={noop} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders ButtonDownloadOptions', () => {
    const { asFragment } = render(<ButtonDownloadOptions onChange={noop} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
