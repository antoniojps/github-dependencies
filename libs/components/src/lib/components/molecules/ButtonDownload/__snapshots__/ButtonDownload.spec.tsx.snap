// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonDownload renders 1`] = `
<DocumentFragment>
  <div
    class="children"
  >
    <div
      class="wrapper"
    >
      <button
        class="button"
      >
        Download
      </button>
      <button
        class="buttonOptions"
      >
        <svg
          fill="none"
          height="14"
          shape-rendering="geometricPrecision"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="14"
        >
          <path
            d="M6 9l6 6 6-6"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ButtonDownload renders ButtonDownloadOptions 1`] = `
<DocumentFragment>
  <div
    class="options"
  >
    <div
      class="name option"
    >
      <label>
        name
      </label>
      <input
        name="name"
        placeholder="dependencies"
        type="text"
        value=""
      />
    </div>
    <div
      class="buttonOptions option"
    >
      <label>
        size
      </label>
      <div
        class="buttons"
      >
        <button
          class="buttonOptionsDisabled"
        >
          1x
        </button>
        <button
          class=""
        >
          2x
        </button>
      </div>
    </div>
    <div
      class="buttonOptions option"
    >
      <label>
        format
      </label>
      <div
        class="buttons"
      >
        <button
          class="buttonOptionsDisabled"
        >
          svg
        </button>
        <button
          class="buttonOptionsDisabled"
        >
          jpg
        </button>
        <button
          class=""
        >
          png
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
